File: synod/src/accounts.rs
//! Synod's own provider accounts: which services it can talk to, and what
//! it authenticates to each of them with.
//!
//! Synod is a desktop application, so keys are typed into a window and
//! kept in this computer's credential manager
//! ([`exarch::provider::keychain`]).  Per account, the credential manager
//! wins — it is the one source a person can see and change from inside
//! synod — with the environment as fallback, resolved and scrubbed by
//! [`CredentialStore::resolve_and_scrub`] exactly as in exarch.  A key
//! from the environment is never silently written into the vault.
//!
//! Which services *exist* is a third thing, and not a secret: the built-in
//! ones are [`exarch::provider::identity::built_in_services`]'s own list,
//! and any further endpoint is declared in
//! `$XDG_CONFIG_HOME/synod/providers.ral`, written by the accounts screen
//! and read by exarch's one declaration decoder ([`exarch::config`]). It
//! holds addresses, never keys.
//!
//! What is provider knowledge — a service's identity, an account's id, the
//! built-in table — lives in [`exarch::provider::accounts`] and
//! [`exarch::provider::identity`]; this module holds only what is about
//! synod's own window: the row a screen draws, where a key is kept, and
//! whether a service can be withdrawn outright.

use exarch::config;
use exarch::provider::accounts::{
    checked_key, declare_endpoint, declared_endpoints, find, withdraw_endpoint,
};
use exarch::provider::credential::{Credential, CredentialStore, NO_AUTH_PLACEHOLDER};
use exarch::provider::identity::{self, Auth, Service};
use exarch::provider::keychain::Keychain;
use exarch::provider::models::{LiveSource, ModelCatalog};
use std::path::PathBuf;
use std::sync::Mutex;

use crate::session::SYNOD;

const DECLARATIONS_FILE: &str = "providers.ral";

/// How the declarations file names itself when it has a complaint to make.
const LABEL: &str = "provider settings";

const KEYCHAIN: Keychain = Keychain::for_app(SYNOD);

fn declarations_path() -> PathBuf {
    SYNOD
        .xdg_dir(ral_core::path::basedir::XdgKind::Config)
        .join(DECLARATIONS_FILE)
}

/// Resolve every account this computer offers, once, at startup.
///
/// The environment sweep first, then the credential manager laid over the
/// top ([`CredentialStore::admit_from`]), so a key typed into the accounts
/// screen outranks a stale variable in the launching environment.
///
/// # Errors
/// Returns `Err` if the declarations file is present but cannot be read or
/// makes no sense; a vault that cannot be reached is not an error here,
/// only fewer accounts to choose from.
///
/// # Panics
/// Must be called while the process is still single-threaded: the
/// credential scrub mutates the environment.
pub fn prepare() -> Result<CredentialStore, String> {
    let declared = config::load_declared(&declarations_path(), LABEL)?;
    let mut store = CredentialStore::resolve_and_scrub(declared);
    store.admit_from(&KEYCHAIN);
    Ok(store)
}

/// Where a row's credential in force actually came from.
///
/// The one fact the screen needs that the shared [`CredentialStore`] does not
/// already carry on its face, since only the store knows which of its doors a
/// key came through.
#[derive(serde::Serialize, Clone, Copy, PartialEq, Eq)]
#[serde(rename_all = "kebab-case")]
pub enum Source {
    /// No credential at all — the row is known but keyless.
    None,
    /// A `ChatGPT` login: the one case a key is never typed, and the sign-in
    /// affordance replaces the key form.
    SignedIn,
    /// The inert bearer a keyless local server is bound to — a state, not a
    /// key anyone typed, so its tail must never be shown as one.
    NoKey,
    Keychain,
    Environment,
}

/// One row of the accounts screen.
#[derive(serde::Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct Account {
    /// The identifier command payloads name this row by — an `AccountId`
    /// rendering, resolved back through [`exarch::provider::accounts::find`].
    /// Never shown; [`Self::label`] is what the screen draws.
    pub id: String,
    /// This account's name among every account currently known —
    /// [`identity::label`], set-relative, so a shared login email still
    /// tells two accounts apart.
    pub label: String,
    pub source: Source,
    /// The last four characters of the key in force, never more.
    pub hint: Option<String>,
    /// The environment variable this account would read from, for the
    /// screen to name when there is no key yet.
    pub env_var: Option<String>,
    /// A declared endpoint's address and protocol; `None` for a built-in
    /// service, whose address is not the user's business.
    pub endpoint: Option<String>,
    pub protocol: Option<String>,
    /// Whether this row's service can be withdrawn outright, rather than
    /// merely having its key taken back — true for a declared endpoint,
    /// never for a built-in one.
    pub withdrawable: bool,
}

/// The accounts screen: every service, whether or not it has a key, and a
/// plain sentence naming where a key typed here would be kept.
#[derive(serde::Serialize, Clone)]
#[serde(rename_all = "camelCase")]
pub struct AccountList {
    pub accounts: Vec<Account>,
    /// "the macOS Keychain", "the Windows Credential Manager", or the
    /// owner-only file a computer with no credential manager falls back to.
    pub vault: String,
    /// The protocols an added endpoint may speak.
    pub protocols: Vec<String>,
}

/// Every account, in the store's own order: built-in services first, then
/// signed-in `ChatGPT` accounts, then declared endpoints.
pub fn list(store: &Mutex<CredentialStore>) -> AccountList {
    let accounts = {
        let store = lock(store);
        let available = store.available();
        store
            .known()
            .iter()
            .map(|account| row(&store, account, &available))
            .collect()
    };
    AccountList {
        accounts,
        vault: KEYCHAIN.vault().to_string(),
        protocols: config::protocols()
            .into_iter()
            .map(str::to_string)
            .collect(),
    }
}

fn row(
    store: &CredentialStore,
    account: &identity::Account,
    available: &[identity::Account],
) -> Account {
    let credential = store.get(&account.id);
    let source = match credential {
        None => Source::None,
        Some(Credential::OAuth(_)) => Source::SignedIn,
        // The inert bearer a keyless local server is bound to is a state,
        // not a key anyone typed; showing its tail would read as one.
        Some(Credential::ApiKey(key)) if key == NO_AUTH_PLACEHOLDER => Source::NoKey,
        // The store remembers which door each key came through, so drawing
        // this list costs the vault no round trip and no unlock prompt.
        Some(Credential::ApiKey(_)) if store.was_admitted(&account.id) => Source::Keychain,
        Some(Credential::ApiKey(_)) => Source::Environment,
    };
    // A built-in service's address is baked in and not the user's business;
    // only a declared endpoint's is shown.
    let withdrawable = identity::built_in(&account.service.name).is_none();
    let (endpoint, protocol) = if withdrawable {
        (
            account.service.endpoint.clone(),
            config::protocol_for_adapter(account.service.adapter).map(str::to_string),
        )
    } else {
        (None, None)
    };
    Account {
        id: account.id.as_str().to_string(),
        label: identity::label(account, available),
        source,
        hint: credential.and_then(hint),
        env_var: env_var_of(&account.service),
        endpoint,
        protocol,
        withdrawable,
    }
}

/// The environment variable an account would read a key from, for the
/// screen to name when there is no key yet — `None` for a `ChatGPT` login
/// and for a declaration that names no variable of its own.
fn env_var_of(service: &Service) -> Option<String> {
    match &service.auth {
        Auth::Env(var) => Some(var.clone()),
        Auth::OAuth | Auth::Unnamed => None,
    }
}

/// The last four characters of an API key: enough to recognise which key is
/// set, useless to anyone reading over a shoulder.
///
/// A plan login has none — the user never typed its token and cannot retype
/// it — and neither has the inert bearer of a keyless server, whose tail
/// would read as a key it does not have.
fn hint(credential: &Credential) -> Option<String> {
    let Credential::ApiKey(key) = credential else {
        return None;
    };
    if key == NO_AUTH_PLACEHOLDER {
        return None;
    }
    let tail = key.char_indices().rev().take(4).last()?.0;
    Some(key[tail..].to_string())
}

/// Keep `key` for the account named `id`, and put it to work at once, with
/// no restart.
///
/// # Errors
/// Returns a plain sentence if no such account is known, if the key is
/// blank, or if this computer's credential manager would not keep it.
pub fn set_key(
    store: &Mutex<CredentialStore>,
    catalog: &Mutex<ModelCatalog<LiveSource>>,
    id: &str,
    key: &str,
) -> Result<(), String> {
    let (account, display) = {
        let store = lock(store);
        let account = find(&store, id)?;
        let available = store.available();
        drop(store);
        let display = identity::label(&account, &available);
        (account, display)
    };
    let key = checked_key(&display, key)?;
    KEYCHAIN.store(account.id.as_str(), &key)?;
    let credential = lock(store).admit_key(&account, key);
    lock(catalog).add_credential(account, credential);
    Ok(())
}

/// Take back the key this window put away for the account named `id`.
///
/// The account does not disappear: it is left known but keyless, or bound
/// again to the key the launching environment supplied all along, which the
/// store kept underneath rather than letting the admission destroy.
///
/// Total in every state, so no caller has to ask first: an id the vault
/// never held is not an error, and revealing the environment's key is the
/// same act as unbinding when there is none to reveal.
///
/// # Errors
/// Returns a plain sentence if no such account is known, or if the
/// credential manager would not give the key up.
pub fn forget_key(store: &Mutex<CredentialStore>, id: &str) -> Result<(), String> {
    let account = find(&lock(store), id)?;
    KEYCHAIN.forget(account.id.as_str())?;
    lock(store).forget(&account.id);
    Ok(())
}

/// Declare another service to talk to: a name, an address, the protocol it
/// speaks, and (unless it wants none) a key.
///
/// # Errors
/// Returns a plain sentence if the name is taken, the address is not one,
/// the protocol is not one of [`config::protocols`], or the declarations
/// file or credential manager refused the write.
pub fn add_endpoint(
    store: &Mutex<CredentialStore>,
    catalog: &Mutex<ModelCatalog<LiveSource>>,
    name: &str,
    endpoint: &str,
    protocol: &str,
    key: Option<&str>,
) -> Result<(), String> {
    let service = declare_endpoint(&lock(store), name, endpoint, protocol, LABEL)?;

    // Checked before a line is written: a key refused after the declaration
    // was saved would leave the service in the file and unusable.
    let typed = match key.map(str::trim).filter(|k| !k.is_empty()) {
        Some(key) => Some(checked_key(service.name.as_str(), key)?),
        None => None,
    };

    let mut declared = declared_endpoints(&lock(store));
    declared.push(service.clone());
    declared.sort_by(|a, b| a.name.as_str().cmp(b.name.as_str()));
    config::save_declared(&declarations_path(), &declared, LABEL)?;

    let account = identity::Account::of_service(service);
    // No key at all is a local server that wants none, admitted with the
    // inert bearer exarch already keeps for exactly that.
    let secret = match typed {
        Some(key) => {
            KEYCHAIN.store(account.id.as_str(), &key)?;
            key
        }
        None => NO_AUTH_PLACEHOLDER.to_string(),
    };
    let credential = lock(store).admit_key(&account, secret);
    lock(catalog).add_credential(account, credential);
    Ok(())
}

/// Withdraw a declared endpoint: out of the live store, out of the file, out
/// of the credential manager.
///
/// # Errors
/// Returns a plain sentence if `id` names no known account, if it names a
/// built-in service rather than a declared endpoint (its key can be taken
/// back, but the service itself cannot be removed), or if the file or
/// credential manager refused the write.
pub fn forget_endpoint(store: &Mutex<CredentialStore>, id: &str) -> Result<(), String> {
    let declared = {
        let mut store = lock(store);
        withdraw_endpoint(&mut store, id)?;
        declared_endpoints(&store)
    };
    config::save_declared(&declarations_path(), &declared, LABEL)?;
    // `id` is the account id's own rendering — `withdraw_endpoint` resolved
    // it — and the vault entry is named by exactly that string.
    KEYCHAIN.forget(id)
}

fn lock<T>(m: &Mutex<T>) -> std::sync::MutexGuard<'_, T> {
    m.lock().unwrap_or_else(std::sync::PoisonError::into_inner)
}

File: synod/ui/js/accounts.js
import { $, invoke, show } from "./core.js";
import { showError, clearError } from "./start-screen.js";
import { actionButton } from "./ui-helpers.js";

// ---- Accounts screen -------------------------------------------------
//
// A key travels one way only: what comes back from the backend is its
// last four characters and where it is kept, so no stored key is ever
// put into an input here.

export async function loadAccounts() {
  try {
    renderAccounts(await invoke("list_accounts"));
  } catch (err) {
    showError("accounts-error", String(err));
  }
}

// Every command that changes something answers with the whole list as
// it now stands, so the screen redraws from what is true rather than
// from what it hoped. `true` means the change went through.
async function accountCommand(command, args) {
  clearError("accounts-error");
  try {
    renderAccounts(await invoke(command, args));
    return true;
  } catch (err) {
    showError("accounts-error", String(err));
    return false;
  }
}

function renderAccounts(list) {
  const accounts = (list && list.accounts) || [];
  const vault = (list && list.vault) || "";
  $("accounts-vault").textContent = vault
    ? "Keys are kept in " + vault + ". Once saved, a key is never shown here again."
    : "";

  const holder = $("accounts-list");
  holder.innerHTML = "";
  if (accounts.length === 0) {
    const empty = document.createElement("p");
    empty.className = "accounts-empty";
    empty.textContent = "No services are set up on this computer yet.";
    holder.appendChild(empty);
  }
  accounts.forEach((account, i) => holder.appendChild(accountRow(account, i)));

  // The plan's card stands whether or not anyone is signed in; it says
  // which of the two rather than coming and going.
  $("accounts-plan-note").textContent = accounts.some((a) => a.source === "signed_in")
    ? "You are signed in with your ChatGPT plan. Signing in again replaces it."
    : "If you pay for ChatGPT, you can use that plan here instead of a key.";

  // A part-filled add form keeps whatever the user had already chosen.
  const select = /** @type {HTMLSelectElement} */ ($("add-protocol"));
  const chosen = select.value;
  const protocols = (list && list.protocols) || [];
  select.innerHTML = "";
  for (const protocol of protocols) {
    const option = document.createElement("option");
    option.value = protocol;
    option.textContent = protocol;
    select.appendChild(option);
  }
  if (protocols.includes(chosen)) select.value = chosen;
}

function accountRow(account, i) {
  const row = document.createElement("div");
  row.className = "account";

  const head = document.createElement("div");
  head.className = "account-head";
  const name = document.createElement("h2");
  name.className = "account-name";
  name.textContent = account.label;
  const state = document.createElement("span");
  state.className = "account-state";
  state.textContent = keyState(account);
  head.append(name, state);
  row.appendChild(head);

  for (const line of accountNotes(account)) {
    const note = document.createElement("p");
    note.className = "account-note";
    note.textContent = line;
    row.appendChild(note);
  }

  // Every account that authenticates with a key gets the key input —
  // a declared endpoint included, since a key rotated at the far end
  // has to be retypeable at this one.
  if (account.source !== "signed_in" && account.source !== "no_key") {
    row.appendChild(keyForm(account, i));
  }
  if (account.withdrawable) {
    const actions = document.createElement("div");
    actions.className = "row-actions";
    actions.appendChild(actionButton("Remove", "btn-mini", () =>
      accountCommand("forget_endpoint", { account: account.id })));
    row.appendChild(actions);
  }
  return row;
}

// "Used without a key" is a state the backend states outright, never
// inferred from a missing hint: a service that wants no key must not
// be read as one still waiting for one.
function keyState(account) {
  if (account.source === "signed_in") return "Signed in";
  if (account.source === "no_key") return "Used without a key";
  return account.hint ? "•••• " + account.hint : "No key yet";
}

function accountNotes(account) {
  if (account.source === "signed_in") {
    return ["You are signed in with your ChatGPT plan, so this one needs no key."];
  }
  if (account.withdrawable) {
    const notes = [];
    if (account.endpoint) notes.push("Address: " + account.endpoint);
    if (account.protocol) notes.push("Way of speaking: " + account.protocol);
    if (account.source === "no_key") {
      notes.push("This one asks for no key, so there is none to type.");
    }
    return notes;
  }
  if (account.source === "no_key") {
    return ["This one asks for no key, so there is none to type."];
  }
  if (account.source === "environment") {
    return [
      "This key came from the environment synod was started in"
        + (account.env_var ? ", as " + account.env_var : "")
        + ". It cannot be changed or removed from this window, but a key typed"
        + " here is used instead of it.",
    ];
  }
  if (account.source === "none") {
    return ["Paste the key this service gave you to start using it."];
  }
  return [];
}

// The one input a key is ever typed into: it starts empty and is thrown
// away with the card on the redraw a save brings, so a saved key is
// never left sitting in the window.
function keyForm(account, i) {
  const form = document.createElement("form");
  form.className = "account-form";
  const inputId = "account-key-" + i;

  const label = document.createElement("label");
  label.className = "field-label";
  label.htmlFor = inputId;
  label.textContent = account.hint
    ? "Replace the key for " + account.label
    : "Key for " + account.label;

  const input = document.createElement("input");
  input.type = "password";
  input.id = inputId;
  input.className = "text-input";
  input.autocomplete = "off";
  input.spellcheck = false;
  input.placeholder = "Paste the key here";

  const actions = document.createElement("div");
  actions.className = "row-actions";
  const save = document.createElement("button");
  save.type = "submit";
  save.className = "btn-mini keep";
  save.textContent = "Save";
  actions.appendChild(save);
  // Only a key this window put away can this window take back.
  if (account.source === "keychain") {
    actions.appendChild(actionButton("Forget", "btn-mini", () =>
      accountCommand("forget_key", { account: account.id })));
  }

  form.append(label, input, actions);
  form.addEventListener("submit", async (event) => {
    event.preventDefault();
    save.disabled = true;
    const saved = await accountCommand("save_key", { account: account.id, key: input.value });
    // A save redraws the whole list, this card with it; a refusal leaves
    // the typed key where it is, to be corrected rather than retyped.
    if (!saved) save.disabled = false;
  });
  return form;
}

$("add-service").addEventListener("submit", async (event) => {
  event.preventDefault();
  const key = /** @type {HTMLInputElement} */ ($("add-key")).value.trim();
  /** @type {HTMLButtonElement} */ ($("add-save")).disabled = true;
  const added = await accountCommand("save_endpoint", {
    name: /** @type {HTMLInputElement} */ ($("add-name")).value,
    endpoint: /** @type {HTMLInputElement} */ ($("add-address")).value,
    protocol: /** @type {HTMLSelectElement} */ ($("add-protocol")).value,
    key: key === "" ? null : key,
  });
  /** @type {HTMLButtonElement} */ ($("add-save")).disabled = false;
  if (added) {
    for (const id of ["add-name", "add-address", "add-key"]) {
      /** @type {HTMLInputElement} */ ($(id)).value = "";
    }
  }
});

$("open-accounts").addEventListener("click", () => {
  clearError("accounts-error");
  show("accounts");
  loadAccounts();
});

$("accounts-done").addEventListener("click", () => show("start"));
